packages/abi/src/parser/specifications/v1/parser.ts

```diff
@@ -0,0 +1,76 @@
+import { FuelError } from '@fuel-ts/errors';
+
+import type { Abi } from '../../abi';
+
+import { mapAttribute } from './map-attribute';
+import { ResolvableType } from './resolvable-type';
+import { ResolvedType } from './resolved-type';
+import type {
+  AbiConfigurableV1,
+  AbiFunctionInputV1,
+  AbiFunctionV1,
+  AbiLoggedTypeV1,
+  AbiMessageTypeV1,
+  AbiSpecificationV1,
+} from './specification';
+
+export class AbiParserV1 {
+  static parse(abi: AbiSpecificationV1): Abi {
+    const resolvableTypes = abi.metadataTypes
+      .map((metadataType) => new ResolvableType(abi, metadataType.metadataTypeId, undefined))
+      .filter(
+        (x) => x.type !== 'struct std::vec::RawVec' && x.type !== 'struct std::bytes::RawBytes'
+      );
+
+    const types = abi.concreteTypes.map((ct) => {
+      const resolvableType = resolvableTypes.find((rt) => rt.metadataTypeId === ct.metadataTypeId);
+
+      const resolvedType = resolvableType
+        ? resolvableType.resolve(ct)
+        : new ResolvedType({ type: ct.type, typeId: ct.concreteTypeId });
+
+      return resolvedType.toAbiType();
+    });
+
+    const getType = (typeId: string | number) => {
+      const type = types.find((t) => t.concreteTypeId === typeId);
+      if (type === undefined) {
+        throw new FuelError(
+          FuelError.CODES.TYPE_ID_NOT_FOUND,
+          `Type with typeId ${typeId} not found`
+        );
+      }
+      return type;
+    };
+
+    return {
+      metadataTypes: resolvableTypes.map((rt) => rt.toAbiType()),
+      types,
+      specVersion: abi.specVersion,
+      encodingVersion: abi.encodingVersion,
+      programType: abi.programType,
+      functions: abi.functions.map((fn: AbiFunctionV1) => ({
+        attributes: fn.attributes?.map(mapAttribute) ?? undefined,
```

Comment:
If these `attributes` are malformed would this throw? I am just speculating...

Reply:
I suppose it could throw - providing the `attributes` are malformed.

Reply:
The `mapAttribute` handles all attributes covered by the spec so it will handle all attributes outputted by `forc`. The `assertUnreachable` at the end will throw and that's okay because it'd come across an unexpected value.